Object Hand Over Action

af-a opened the pull request:
# Summary of changes

This pull request contains an implementation of a basic hand-over action used in the research and development project titled: "Incorporating Contextual Knowledge Into Human-Robot Collaborative Task Execution". Please refer to this [technical report](https://pub.h-brs.de/frontdoor/index/index/searchtype/latest/docId/4828/start/9/rows/25) based on the project for more details.

[This video](https://www.youtube.com/watch?v=GguFJ2a7O6E) demonstrates the action in action :).

# TODO before merging

* [x] Remove unnecessary files
* [x] Test and validate implementation
* [x] Clean up code if necessary
* [x] Implement move_base feature to approach person before handing object over
* [x] Implement smarter detection of object receiver: picking the nearest detected person

af-a committed:
[hand_over_action] Add initial action implementation and parameter files

alex-mitrevski committed:
Added a demo for context-aware item hand over

alex-mitrevski committed:
[demo_context_aware_hand_over/states/hand_over] Fixed incorrect action server name

alex-mitrevski committed:
[demo_context_aware_hand_over/sm] Updated the bb ratios for the different contexts

af-a committed:
[hand_over_action] Correct dmp weight files (RPY vector sizes)

af-a committed:
[hand_over_action] Change policy param pickle files to python 2 format

af-a committed:
[hand_over_action] Fix action params: config file paths

af-a committed:
[hand_over_action] Change argument parsing in client test file

af-a committed:
[hand_over_action] Add force sensing for object reception (testing)

af-a committed:
Merge remote-tracking branch 'origin/feature/hand-over-action' into feature/hand-over-action

af-a committed:
[hand_over_action] Implement change detection for object reception

af-a committed:
[demo_context_aware_hand_over] Add Alex's fixes to scenario files

af-a committed:
[hand_over_action] Extend object reception to check both dims: x and z

af-a committed:
[hand_over_action] Add choice for context-awareness in action & scenario

af-a committed:
[hand_over_action] Remove explicit vocalization of detected posture

mas-hsr committed:
[hand_over_action] Minor modifications (debugging outputs, fixes, parameter tweaking)

mas-hsr committed:
[context_aware_hand_over/identify_posture] Adapted the Person msg use to the new msg format

mas-hsr committed:
[context_aware_hand_over/identify_posture] The object is now handed over to the closest person

af-a committed:
[hand-over-action] Remove unnecessary learned policy parameter files

af-a committed:
[hand-over-action] Add boolean release detection parameter, and some code clean-up

af-a committed:
Merge branch 'feature/hand-over-action' of github.com:b-it-bots/mas_domestic_robotics into feature/hand-over-action

af-a committed:
[hand-over-action] Fix incorrect release_detection variable naming

af-a committed:
[hand-over-action] Fix incorrect result when object is released without detection

mas-hsr committed:
[demo_context_aware_hand_over] Included the newly added 'release_detection' action parameter

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_manipulation_actions/mdr_hand_over_action/README.md:
```diff
@@ -0,0 +1,3 @@
+# mdr_handle_open_action
+
+An action for handing a grasped object over to a person.
```
It would be great to expand on the documentation (e.g. following the model of the [pickup action](https://github.com/b-it-bots/mas_domestic_robotics/tree/devel/mdr_planning/mdr_actions/mdr_manipulation_actions/mdr_pickup_action))

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_perception_actions/mdr_find_people/ros/src/mdr_find_people/action_states.py:
```diff
@@ -48,7 +48,7 @@ def __init__(self):
 
         self._listener = tf.TransformListener()
-        self.pointcloud_topic = rospy.get_param("~pointcloud_topic", '/rectified_points')
+        self.pointcloud_topic = rospy.get_param("~pointcloud_topic", '/hsrb/head_rgbd_sensor/depth_registered/rectified_points')
```
This line should be reverted back; the parameter will come from an HSR-specific launch file.

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_manipulation_actions/mdr_hand_over_action/ros/launch/hand_over.launch:
```diff
@@ -0,0 +1,13 @@
+<launch>
+    <node pkg="mdr_hand_over_action" type="hand_over_action" name="hand_over_server" output="screen">
+        <param name="move_arm_server" value="move_arm_server" />
+        <param name="gripper_controller_pkg_name" value="mas_hsr_gripper_controller" />
```
The value should be changed to `mdr_gripper_controller`. We'll have an HSR-specific launch file to set this for the HSR.

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_manipulation_actions/mdr_hand_over_action/ros/scripts/hand_over_action:
```diff
@@ -0,0 +1,58 @@
+    The server expects the following parameters to be made available on the ROS parameter server:
+    '''
+    def __init__(self):
+        gripper_controller_pkg_name = rospy.get_param('~gripper_controller_pkg_name', 'mas_hsr_gripper_controller')
```
The default value here should be `mdr_gripper_controller`.

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_manipulation_actions/mdr_hand_over_action/ros/src/mdr_hand_over_action/action_states.py:
```diff
@@ -0,0 +1,227 @@
+
+class HandOverSM(ActionSMBase):
+    def __init__(self, timeout=120.0,
+                 gripper_controller_pkg_name='mas_hsr_gripper_controller',
```
Default value `mdr_gripper_controller`

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_manipulation_actions/mdr_hand_over_action/ros/src/mdr_hand_over_action/action_states.py:
```diff
@@ -0,0 +1,227 @@
+class HandOverSM(ActionSMBase):
+    def __init__(self, timeout=120.0,
+                 gripper_controller_pkg_name='mas_hsr_gripper_controller',
+                 force_sensor_topic='/hsrb/wrist_wrench/raw',
```
Use a robot-independent default value, e.g. `/wrench/raw`

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_perception_actions/mdr_find_people/ros/launch/find_people.launch:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0"?>
 <launch>
-    <arg name="pointcloud_topic" default="/mdr_perception/rectified_points"/>
+    <arg name="pointcloud_topic" default="/hsrb/head_rgbd_sensor/depth_registered/rectified_points"/>
```
This line should be reverted back; the parameter will come from an HSR-specific launch file.

af-a committed:
[hand_over_action] Switch to robot-independent topics and packages

Requested review changes

af-a committed:
[hand_over_action] Update README with more detailed description

sthoduka committed:
[hand_over_action] move arm to neutral position in init state (otherwise the execution of the DMP is weird); fix readme

alex-mitrevski committed:
Merge branch 'devel' into feature/hand-over-action

af-a committed:
Add point cloud filtering to improve person pose estimation

af-a committed:
Utilize person pose for pre-approach (in action and scenario)

af-a committed:
[demo_hand_over] Add transition to detect_people if identify_posture fails

alex-mitrevski commented on mdr_planning/mdr_actions/mdr_manipulation_actions/mdr_hand_over_action/ros/launch/hand_over.launch:
```diff
@@ -2,12 +2,14 @@
         <param name="move_arm_server" value="move_arm_server" />
-        <param name="gripper_controller_pkg_name" value="mdr_gripper_controller" />
+        <param name="move_base_server" value="move_base_server" />
+        <param name="gripper_controller_pkg_name" value="mas_hsr_gripper_controller" />
```
Leave `mdr_gripper_controller` here please (`mas_hsr_gripper_controller` in the launcher in `mas_hsr`).

alex-mitrevski commented on mdr_planning/mdr_scenarios/mdr_demos/mdr_demo_context_aware_hand_over/ros/launch/hand_over.launch:
```diff
@@ -9,7 +9,7 @@
 
     <!-- Actions -->
-    <include file="$(find mdr_find_people)/ros/launch/find_people.launch" />
+    <include file="$(find mas_hsr_find_people)/ros/launch/find_people.launch" />
```
Same as above.

alex-mitrevski committed:
[hand_over_action] Fixed a small bug due to which the arm wasn't moving to the hand over pose

In particular, the timestamp of the hand over pose was set to the
current ROS time, so the pose could not be transformed during DMP
execution

af-a committed:
[hand_over_action] Add force sensor topic

af-a committed:
[hand_over_action] Update action client script

af-a committed:
[hand_over_action] Modify release detection strategy to use force readings in z-axis

af-a committed:
[hand_over_action] Remove hsr-specific topics

af-a committed:
[hand_over_action] Remove specification of post message timestamp

af-a committed:
[hand_over_action] Remove unnecessary comments and add move_base goal cancellation after motion duration

af-a committed:
Merge branch 'feature/hand-over-action' of github.com:b-it-bots/mas_domestic_robotics into feature/hand-over-action

alex-mitrevski committed:
Merge branch 'devel' of github.com:b-it-bots/mas_domestic_robotics into feature/hand-over-action

alex-mitrevski committed:
[hand_over_action/action_states] Resolved some Codacy complaints

alex-mitrevski committed:
[find_people/src/find_people] Removed unnecessary line
